server/src/main/java/org/elasticsearch/index/shard/RemoveCorruptedSegmentsCommand.java

```diff
@@ -0,0 +1,320 @@
+                    checker.setChecksumsOnly(true);
+                    checker.setInfoStream(printStream, terminal.isPrintable(Terminal.Verbosity.VERBOSE));
+
+                    status = checker.checkIndex(null);
+
+                    if (doDryRun == true) {
+                        if (status.missingSegments == false) {
+                            if (status.clean == false) {
+                                terminal.println("WARNING: Corrupted segments found - " + status.totLoseDocCount
+                                    + " documents are damaged");
+                            } else {
+                                terminal.println("Index is clean");
+                            }
+                        } else {
+                            terminal.println("WARNING: There are missing segments");
+                        }
+                        return;
+                    } else {
+                        if (status.missingSegments == false) {
+                            if (status.clean == false) {
+                                warnMessage(terminal, status);
+
+                                terminal.println("Writing...");
+                                checker.exorciseIndex(status);
+
+                                terminal.println("OK");
+                                terminal.println("Wrote new segments file \"" + status.segmentsFileName + "\"");
+                            } else {
+                                terminal.println("Index is clean");
+                            }
+                            dropCorruptMarkerFiles(terminal, directory);
+                        } else {
+                            // TODO: drop it when creating new empty commit is fixed
+                            // throw new UserException(FAILURE, "There are missing segments");
+                        }
+                    }
+                } catch (LockObtainFailedException lofe) {
+                    throw new UserException(MISCONFIGURATION,
+                        "Failed to lock shard's directory at [" + path + "], is Elasticsearch still running?");
+                }
+
+                // IndexWriter acquires directory lock by its own
+                if (status.missingSegments) {
+                    adjustMissedSegments(terminal, dir, path);
+
+                    dropCorruptMarkerFiles(terminal, directory);
+                }
+            }
+        }
+    }
+
+    private void adjustMissedSegments(Terminal terminal, Directory dir, Path path) throws IOException {
+        terminal.println("There are missing segments");
+
+        terminal.println("WARNING! It drops all Lucene index files");
+
+        confirm("Continue and clean up Lucene index files ?", terminal);
+        // TODO: waiting for Lucene fixes or IndexWriter (to be able to create a new segment while other files there)
+        // or wait while Lucene CheckIndex makes files non corrupted
+        Lucene.cleanLuceneIndex(dir);
```

Comment:
I just had a chat with @vladimirdolzhenko and dropping the whole index is probably the safest option at this point when CheckIndex.Status.missing is `true`

Reply:
I added a link to [LUCENE-6762](https://issues.apache.org/jira/browse/LUCENE-6762) to the comment